File: box-and-box/ports/python/strategic.py
"""strategic.py - Strategic / Coalitional Arithmetic, faithful Python port (v0.7)

Over a CONCURRENT GAME STRUCTURE - states, agents, a set of moves per agent per
state, and a transition that consumes one move from every agent - "coalition C
can ensure phi" means C has a joint strategy such that, whatever the others do,
phi results. One step is effectivity [C]O phi; over time, <<C>>[]phi (maintain -
greatest fixpoint) and <<C>><>phi (reach - least fixpoint), both built from the
CONTROLLABLE PREDECESSOR. Laws S1-S8, SB1-SB3. Mirrors strategic.mjs.

States are dicts compared by object identity (mirroring JS .includes / indexOf).
"""


def atom(name, pred):
    return {"t": "atom", "name": name, "pred": pred}


def not_(a):
    return {"t": "not", "a": a}


def and_(a, b):
    return {"t": "and", "a": a, "b": b}


def or_(a, b):
    return {"t": "or", "a": a, "b": b}


def holds(f, s):
    t = f["t"]
    if t == "atom":
        return bool(f["pred"](s))
    if t == "not":
        return not holds(f["a"], s)
    if t == "and":
        return holds(f["a"], s) and holds(f["b"], s)
    if t == "or":
        return holds(f["a"], s) or holds(f["b"], s)
    return False


TOP = or_(atom("\u22a4", lambda s: True), not_(atom("\u22a4", lambda s: True)))
BOT = and_(atom("\u22a5", lambda s: False), not_(atom("\u22a5", lambda s: False)))


def Game(states, agents, moves, delta):
    return {"states": states, "agents": agents, "moves": moves, "delta": delta}


def others(model, C):
    return [a for a in model["agents"] if a not in C]


def _product(model, agents, state):
    """Cartesian product of agents' move sets at a state -> list of {agent: moveId}."""
    acc = [{}]
    for a in agents:
        ms = model["moves"](a, state)
        nx = []
        for p in acc:
            for m in ms:
                d = dict(p)
                d[a] = m
                nx.append(d)
        acc = nx
    return acc


def _contains(seq, s):
    return any(s is x for x in seq)


def _force1(model, C, state, in_set):
    """Controllable predecessor: exists moves for C, forall moves for the rest, succ in set."""
    cm = _product(model, C, state)
    om = _product(model, others(model, C), state)
    for c in cm:
        ok = True
        for o in om:
            jm = dict(c)
            jm.update(o)
            if not in_set(model["delta"](state, jm)):
                ok = False
                break
        if ok:
            return True
    return False


def effectivity(model, C, state, f):
    return _force1(model, C, state, lambda s: holds(f, s))  # [C]O f at a state


def can_ensure_next(model, C, f):
    return [q for q in model["states"] if effectivity(model, C, q, f)]


def can_maintain(model, C, f):
    """<<C>>[]f - greatest fixpoint nuW.(f & Pre_C W)."""
    W = [s for s in model["states"] if holds(f, s)]
    while True:
        cur = W
        W2 = [q for q in cur if _force1(model, C, q, lambda s: _contains(cur, s))]
        if len(W2) == len(W):
            return W2
        W = W2


def can_reach(model, C, f):
    """<<C>><>f - least fixpoint muW.(f | Pre_C W)."""
    W = [s for s in model["states"] if holds(f, s)]
    while True:
        cur = W
        add = [q for q in model["states"]
               if not _contains(cur, q) and _force1(model, C, q, lambda s: _contains(cur, s))]
        if not add:
            return W
        W = W + add


def can_until(model, C, f, g):
    """<<C>>(f U g) - least fixpoint muW.(g | (f & Pre_C W))."""
    W = [s for s in model["states"] if holds(g, s)]
    while True:
        cur = W
        add = [q for q in model["states"]
               if not _contains(cur, q) and holds(f, q) and _force1(model, C, q, lambda s: _contains(cur, s))]
        if not add:
            return W
        W = W + add


def can_ensure(model, C, f, q):
    return _contains(can_reach(model, C, f), q)  # <<C>><>f from q


def can_keep(model, C, f, q):
    return _contains(can_maintain(model, C, f), q)  # <<C>>[]f from q


def oblige(model, C, f, q):
    """ought-implies-can: dischargeable only if C can ensure it."""
    return "discharge" if can_ensure(model, C, f, q) else "escalate"


def executable(model, C, f, q, common_knowledge):
    """coordination is executable only with ability AND common knowledge of the plan."""
    return can_ensure(model, C, f, q) and bool(common_knowledge)

File: witness/src/AmpersandBoxDesign/box-and-box/strategic.mjs
// strategic.mjs — Strategic / Coalitional Arithmetic, faithful runtime (v0.7)
// The seventh and last rung: ability. Over a CONCURRENT GAME STRUCTURE — states, agents,
// a set of moves per agent per state, and a transition that consumes one move from every
// agent — "coalition C can ensure φ" means C has a joint strategy such that, whatever the
// other agents do, φ results. One step is effectivity [C]◯φ; over time, ⟨⟨C⟩⟩□φ (maintain —
// a greatest fixpoint) and ⟨⟨C⟩⟩◊φ (reach — a least fixpoint), both built from the
// CONTROLLABLE PREDECESSOR. Laws S1–S8. Grounded in Pauly 2002 (coalition logic) and
// Alur, Henzinger & Kupferman 2002 (Alternating-time Temporal Logic).

// ---- propositional formulas over states ----
export const atom = (name, pred) => ({ t: 'atom', name, pred });
export const not = (a) => ({ t: 'not', a });
export const and = (a, b) => ({ t: 'and', a, b });
export const or = (a, b) => ({ t: 'or', a, b });
export function holds(f, s) {
  switch (f.t) {
    case 'atom': return !!f.pred(s);
    case 'not': return !holds(f.a, s);
    case 'and': return holds(f.a, s) && holds(f.b, s);
    case 'or': return holds(f.a, s) || holds(f.b, s);
    default: return false;
  }
}
export const TOP = or(atom('⊤', () => true), not(atom('⊤', () => true)));
export const BOT = and(atom('⊥', () => false), not(atom('⊥', () => false)));

// ---- game: states, agents, moves(agent,state)→moveId[], delta(state,{agent:moveId})→state ----
export const Game = ({ states, agents, moves, delta }) => ({ states, agents, moves, delta });
export const others = (model, C) => model.agents.filter((a) => !C.includes(a));

// cartesian product of the agents' move sets at a state → list of {agent: moveId}
function product(model, agents, state) {
  let acc = [{}];
  for (const a of agents) { const ms = model.moves(a, state); const nx = [];
    for (const p of acc) for (const m of ms) nx.push({ ...p, [a]: m }); acc = nx; }
  return acc;
}
// controllable predecessor at a state: ∃ moves for C, ∀ moves for the rest, successor ∈ set
function force1(model, C, state, inSet) {
  const cm = product(model, C, state), om = product(model, others(model, C), state);
  return cm.some((c) => om.every((o) => inSet(model.delta(state, { ...c, ...o }))));
}

// ---- the operators ----
export const effectivity = (model, C, state, f) => force1(model, C, state, (s) => holds(f, s)); // [C]◯f at a state
export const canEnsureNext = (model, C, f) => model.states.filter((q) => effectivity(model, C, q, f));

// ⟨⟨C⟩⟩□f — greatest fixpoint νW.(f ∧ Pre_C W): the states from which C can keep f forever
export function canMaintain(model, C, f) {
  const phi = (s) => holds(f, s); let W = model.states.filter(phi);
  for (;;) { const inW = (s) => W.includes(s); const W2 = W.filter((q) => force1(model, C, q, inW));
    if (W2.length === W.length) return W2; W = W2; }
}
// ⟨⟨C⟩⟩◊f — least fixpoint μW.(f ∨ Pre_C W): the states from which C can drive the system to f
export function canReach(model, C, f) {
  const phi = (s) => holds(f, s); let W = model.states.filter(phi);
  for (;;) { const inW = (s) => W.includes(s);
    const add = model.states.filter((q) => !inW(q) && force1(model, C, q, inW));
    if (!add.length) return W; W = W.concat(add); }
}
// ⟨⟨C⟩⟩(f U g) — least fixpoint μW.(g ∨ (f ∧ Pre_C W))
export function canUntil(model, C, f, g) {
  const phi = (s) => holds(f, s), psi = (s) => holds(g, s); let W = model.states.filter(psi);
  for (;;) { const inW = (s) => W.includes(s);
    const add = model.states.filter((q) => !inW(q) && phi(q) && force1(model, C, q, inW));
    if (!add.length) return W; W = W.concat(add); }
}

// convenience: ability from a particular state
export const canEnsure = (model, C, f, q) => canReach(model, C, f).includes(q); // ⟨⟨C⟩⟩◊f from q
export const canKeep = (model, C, f, q) => canMaintain(model, C, f).includes(q); // ⟨⟨C⟩⟩□f from q
// ought-implies-can: an obligation to ensure f is dischargeable only if C can ensure it
export const oblige = (model, C, f, q) => canEnsure(model, C, f, q) ? 'discharge' : 'escalate';
// coordination is executable only with ability AND common knowledge of the plan (the epistemic link)
export const executable = (model, C, f, q, commonKnowledge) => canEnsure(model, C, f, q) && !!commonKnowledge;

export default {
  atom, not, and, or, holds, TOP, BOT, Game, others, effectivity, canEnsureNext,
  canMaintain, canReach, canUntil, canEnsure, canKeep, oblige, executable
};
